docs/resources/server/recipes/custom.md

```diff
@@ -0,0 +1,699 @@
+    ).apply(inst, RightClickBlockRecipe::new));
+    public static final StreamCodec<RegistryFriendlyByteBuf, RightClickBlockRecipe> STREAM_CODEC =
+            StreamCodec.composite(
+                    ByteBufCodecs.idMapper(Block.BLOCK_STATE_REGISTRY), RightClickBlockRecipe::getInputState,
+                    Ingredient.CONTENTS_STREAM_CODEC, RightClickBlockRecipe::getInputItem,
+                    ItemStack.STREAM_CODEC, RightClickBlockRecipe::getResult,
+                    RightClickBlockRecipe::new
+            );
+
+    // Return our map codec.
+    @Override
+    public MapCodec<RightClickBlockRecipe> codec() {
+        return CODEC;
+    }
+
+    // Return our stream codec.
+    @Override
+    public StreamCodec<RegistryFriendlyByteBuf, RightClickBlockRecipe> streamCodec() {
+        return STREAM_CODEC;
+    }
+}
+```
+
+Like with the type, we register our serializer:
+
+```java
+public static final DeferredRegister<RecipeType<?>> RECIPE_SERIALIZERS =
+        DeferredRegister.create(Registries.RECIPE_SERIALIZER, ExampleMod.MOD_ID);
+
+public static final Supplier<RecipeSerializer<RightClickBlockRecipe>> RIGHT_CLICK_BLOCK =
+        RECIPE_SERIALIZERS.register("right_click_block", RightClickBlockRecipeSerializer::new);
+```
+
+And similarly, we must also override `#getSerializer` in our recipe, like so:
+
+```java
+public class RightClickBlockRecipe implements Recipe<RightClickBlockInput> {
+    // other stuff here
+
+    @Override
+    public RecipeSerializer<? extends Recipe<RightClickBlockInput>> getSerializer() {
+        return RIGHT_CLICK_BLOCK.get();
+    }
+}
+```
+
+## The Crafting Mechanic
+
+Now that all parts of your recipe are complete, you can make yourself some recipe JSONs (see the [datagen] section for that) and then query the recipe manager for your recipes, like above. What you then do with the recipe is up to you. A common use case would be a machine that can process your recipes, storing the active recipe as a field.
+
+In our case, however, we want to apply the recipe when an item is right-clicked on a block. We will do so using an [event handler][event]. Keep in mind that this is an example implementation, and you can alter this in any way you like (so long as you run it on the server). As we want the interaction state to match on both the client and server, we will also need to [sync any relevant input states across the network][networking].
+
+We can set up a simple network implementation to sync the recipe inputs like so:
+
+```java
+// A basic packet class, must be registered.
+public record ClientboundRightClickBlockRecipesPayload(
+    Set<BlockState> inputStates, Set<Holder<Item>> inputItems
+) implements CustomPacketPayload {
+
+    // ...
+}
+
+
+// Packet stores data in an instance class.
+// Present on both server and client to do initial matching.
+// Resource listener so it can be reloaded when recipes are.
+public class RightClickBlockRecipeInputs extends SimplePreparableReloadListener<Void> {
```

Comment:
A `ResourceManagerReloadListener` would be more concise as it is effectively a reload listener with only an `apply()` method and no state that's passed between the two phases.